dicom_utils/volume.py

```diff
@@ -252,29 +252,99 @@ def get_indices(self, total_frames: Optional[int]) -> Tuple[int, Optional[int],
     ):
         self.reduction = reduction
+        self.output_frames = output_frames
+        self.skip_edge_frames = skip_edge_frames
+
+    def __call__(self, x: Union[T, U], use_nvjpeg: Optional[bool] = None, **kwargs) -> Union[T, U]:
+        if isinstance(x, Dataset):
+            return self.handle_dicom(x, use_nvjpeg=use_nvjpeg, **kwargs)
+        else:
+            return self.handle_array(x)
 
     def get_indices(self, total_frames: Optional[int]) -> Tuple[int, int, int]:
         raise NotImplementedError
 
-    def handle_dicom(self, dcm: U) -> U:
-        # slice dicom at index 0 and read the associated numpy array
-        sliced = self.slice_dicom(dcm, 0, 1, 1)
-        arr = sliced.pixel_array
-
-        # iterate through remaining slices, applying reduction in-place
-        for i in range(1, dcm.NumberOfFrames):
-            other = self.slice_dicom(dcm, i, i + 1, 1).pixel_array
-            arr = self.reduction(arr, other, out=arr)
-        dcm = self.update_pixel_data(sliced, [arr.tobytes()], preserve_compression=False)
-        dcm.pixel_array
+    def handle_dicom(self, dcm: U, use_nvjpeg: Optional[bool] = None, **kwargs) -> U:
+        chunks = [chunk.tobytes() for chunk in self.iterate_chunks(dcm, self.reduction, use_nvjpeg, **kwargs)]
+        if not chunks:
+            raise RuntimeError("No frames remain in the sliced DICOM")
+        assert len(chunks) == self.output_frames
+
+        dcm = self.update_pixel_data(dcm, chunks, preserve_compression=False)
+        assert not dcm.file_meta.TransferSyntaxUID.is_compressed, "DICOM should be decompressed after ReduceVolume"
         return dcm
 
     def handle_array(self, x: T) -> T:
         raise NotImplementedError
+
+    def iterate_chunks(
+        self,
+        dcm: Dataset,
+        reduction: Optional[Callable[..., np.ndarray]],
+        use_nvjpeg: Optional[bool] = None,
+        **kwargs,
+    ) -> Iterator[np.ndarray]:
+        r"""Iterates through the DICOM volume, yielding chunks based on the reduction parameters.
+        If a reduction is given, each chunk will be reduced before being yielded.
+        """
+        # TODO: This is a bit of a hack to avoid circular imports. Fixing it will require a larger refactor.
+        from .dicom import decompress
+
+        start = self.skip_edge_frames
+        stop = dcm.NumberOfFrames - self.skip_edge_frames
+        chunk_size = (stop - start) // self.output_frames
+
+        yield_count = 0
+        while start < stop and yield_count < self.output_frames:
+            # If this is the last chunk, make sure it includes all remaining frames
+            is_last_chunk = yield_count == self.output_frames - 1
+            chunk_end = min(start + chunk_size, stop) if not is_last_chunk else stop
+
+            # Slice the chunk and decompress if necessary
+            sliced = self.slice_dicom(dcm, start, chunk_end, stride=1)
+            sliced = decompress(sliced, use_nvjpeg=use_nvjpeg, **kwargs)
```

Comment:
I think we want to decompress slice by slice to save memory, so this GPU decompression should take place in the `for` loop below.